{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " ### TESTING ML MODELS WITH LEAVE-ONE-OUT CROSS VALIDATION\n",
    "* Mollah Lab: CREWdb\n",
    "* Last Updated: 1/13/23"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "id": "DAJWWVo8saf8"
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.model_selection import LeaveOneOut\n",
    "from sklearn.naive_bayes import GaussianNB\n",
    "from sklearn.metrics import mean_squared_error\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.feature_selection import SelectKBest, f_classif\n",
    "from sklearn.pipeline import make_pipeline\n",
    "from sklearn.feature_selection import VarianceThreshold\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.metrics import f1_score\n",
    "from sklearn.metrics import precision_recall_curve\n",
    "from sklearn.metrics import plot_confusion_matrix\n",
    "from sklearn.metrics import confusion_matrix\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Pre-Processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 823
    },
    "id": "qoPz9jXks-PL",
    "outputId": "5154e9a9-a184-4088-dfae-9501ebedd782"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID (REMOVE)</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>REW</th>\n",
       "      <th>HGNC approved name</th>\n",
       "      <th>Pfam domains</th>\n",
       "      <th>HGNC gene family tag</th>\n",
       "      <th>HGNC gene family description</th>\n",
       "      <th>Function</th>\n",
       "      <th>Modification</th>\n",
       "      <th>Protein complex</th>\n",
       "      <th>Target molecule</th>\n",
       "      <th>Target entity</th>\n",
       "      <th>Product</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10389</td>\n",
       "      <td>SCML2</td>\n",
       "      <td>Reader</td>\n",
       "      <td>sex comb on midleg-like 2 (Drosophila)</td>\n",
       "      <td>DUF3588 PF12140 353-468, MBT PF02820 67-139 17...</td>\n",
       "      <td>SAMD</td>\n",
       "      <td>Sterile alpha motif (SAM) domain containing</td>\n",
       "      <td>Polycomb group (PcG) protein</td>\n",
       "      <td>#</td>\n",
       "      <td>PRC1</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22955</td>\n",
       "      <td>SCMH1</td>\n",
       "      <td>Reader</td>\n",
       "      <td>sex comb on midleg homolog 1 (Drosophila)</td>\n",
       "      <td>DUF3588 PF12140 355-469, MBT PF02820 62-134 17...</td>\n",
       "      <td>SAMD</td>\n",
       "      <td>Sterile alpha motif (SAM) domain containing</td>\n",
       "      <td>Polycomb group (PcG) protein</td>\n",
       "      <td>#</td>\n",
       "      <td>PRC1</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5252</td>\n",
       "      <td>PHF1</td>\n",
       "      <td>Reader</td>\n",
       "      <td>PHD finger protein 1</td>\n",
       "      <td>Mtf2_C PF14061 525-566, PHD PF00628 89-142, Pf...</td>\n",
       "      <td>TDRD, PHF</td>\n",
       "      <td>Tudor domain containing, Zinc fingers, PHD-type</td>\n",
       "      <td>Polycomb group (PcG) protein</td>\n",
       "      <td>#</td>\n",
       "      <td>PRC2</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8726</td>\n",
       "      <td>EED</td>\n",
       "      <td>Reader</td>\n",
       "      <td>embryonic ectoderm development</td>\n",
       "      <td>Pfam-B_26038 PB026038 1-39, Pfam-B_6684 PB0066...</td>\n",
       "      <td>WDR</td>\n",
       "      <td>WD repeat domain containing</td>\n",
       "      <td>Polycomb group (PcG) protein</td>\n",
       "      <td>#</td>\n",
       "      <td>PRC2</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>84181</td>\n",
       "      <td>CHD6</td>\n",
       "      <td>Reader</td>\n",
       "      <td>chromodomain helicase DNA binding protein 6</td>\n",
       "      <td>BRK PF07533 2400-2441, Chromo PF00385 292-354 ...</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "      <td>Chromatin remodeling</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "      <td>chromatin</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID (REMOVE) SYMBOL     REW                           HGNC approved name  \\\n",
       "0        10389  SCML2  Reader       sex comb on midleg-like 2 (Drosophila)   \n",
       "1        22955  SCMH1  Reader    sex comb on midleg homolog 1 (Drosophila)   \n",
       "2         5252   PHF1  Reader                         PHD finger protein 1   \n",
       "3         8726    EED  Reader               embryonic ectoderm development   \n",
       "4        84181   CHD6  Reader  chromodomain helicase DNA binding protein 6   \n",
       "\n",
       "                                        Pfam domains HGNC gene family tag  \\\n",
       "0  DUF3588 PF12140 353-468, MBT PF02820 67-139 17...                 SAMD   \n",
       "1  DUF3588 PF12140 355-469, MBT PF02820 62-134 17...                 SAMD   \n",
       "2  Mtf2_C PF14061 525-566, PHD PF00628 89-142, Pf...            TDRD, PHF   \n",
       "3  Pfam-B_26038 PB026038 1-39, Pfam-B_6684 PB0066...                  WDR   \n",
       "4  BRK PF07533 2400-2441, Chromo PF00385 292-354 ...                    #   \n",
       "\n",
       "                      HGNC gene family description  \\\n",
       "0      Sterile alpha motif (SAM) domain containing   \n",
       "1      Sterile alpha motif (SAM) domain containing   \n",
       "2  Tudor domain containing, Zinc fingers, PHD-type   \n",
       "3                      WD repeat domain containing   \n",
       "4                                                #   \n",
       "\n",
       "                       Function Modification Protein complex Target molecule  \\\n",
       "0  Polycomb group (PcG) protein            #            PRC1               #   \n",
       "1  Polycomb group (PcG) protein            #            PRC1               #   \n",
       "2  Polycomb group (PcG) protein            #            PRC2               #   \n",
       "3  Polycomb group (PcG) protein            #            PRC2               #   \n",
       "4          Chromatin remodeling            #               #       chromatin   \n",
       "\n",
       "  Target entity Product  \n",
       "0             #       #  \n",
       "1             #       #  \n",
       "2             #       #  \n",
       "3             #       #  \n",
       "4             #       #  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Reading in bulk data\n",
    "data = pd.read_csv(\"./data/dataset_cleaned_397.csv\").rename(columns={'REW (Convert to Numbers)': 'REW'})\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "seBZH03btIjZ",
    "outputId": "24f32f4f-beda-455e-a4b5-e9fa6a4f5b46"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID (REMOVE)</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>REW</th>\n",
       "      <th>HGNC approved name</th>\n",
       "      <th>Pfam domains</th>\n",
       "      <th>HGNC gene family tag</th>\n",
       "      <th>HGNC gene family description</th>\n",
       "      <th>Function</th>\n",
       "      <th>Modification</th>\n",
       "      <th>Protein complex</th>\n",
       "      <th>Target molecule</th>\n",
       "      <th>Target entity</th>\n",
       "      <th>Product</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10389</td>\n",
       "      <td>SCML2</td>\n",
       "      <td>Reader</td>\n",
       "      <td>sex comb on midleg-like 2 (Drosophila)</td>\n",
       "      <td>DUF3588 PF12140 353-468, MBT PF02820 67-139 17...</td>\n",
       "      <td>SAMD</td>\n",
       "      <td>Sterile alpha motif (SAM) domain containing</td>\n",
       "      <td>Polycomb group (PcG) protein</td>\n",
       "      <td>#</td>\n",
       "      <td>PRC1</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22955</td>\n",
       "      <td>SCMH1</td>\n",
       "      <td>Reader</td>\n",
       "      <td>sex comb on midleg homolog 1 (Drosophila)</td>\n",
       "      <td>DUF3588 PF12140 355-469, MBT PF02820 62-134 17...</td>\n",
       "      <td>SAMD</td>\n",
       "      <td>Sterile alpha motif (SAM) domain containing</td>\n",
       "      <td>Polycomb group (PcG) protein</td>\n",
       "      <td>#</td>\n",
       "      <td>PRC1</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5252</td>\n",
       "      <td>PHF1</td>\n",
       "      <td>Reader</td>\n",
       "      <td>PHD finger protein 1</td>\n",
       "      <td>Mtf2_C PF14061 525-566, PHD PF00628 89-142, Pf...</td>\n",
       "      <td>TDRD, PHF</td>\n",
       "      <td>Tudor domain containing, Zinc fingers, PHD-type</td>\n",
       "      <td>Polycomb group (PcG) protein</td>\n",
       "      <td>#</td>\n",
       "      <td>PRC2</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8726</td>\n",
       "      <td>EED</td>\n",
       "      <td>Reader</td>\n",
       "      <td>embryonic ectoderm development</td>\n",
       "      <td>Pfam-B_26038 PB026038 1-39, Pfam-B_6684 PB0066...</td>\n",
       "      <td>WDR</td>\n",
       "      <td>WD repeat domain containing</td>\n",
       "      <td>Polycomb group (PcG) protein</td>\n",
       "      <td>#</td>\n",
       "      <td>PRC2</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>84181</td>\n",
       "      <td>CHD6</td>\n",
       "      <td>Reader</td>\n",
       "      <td>chromodomain helicase DNA binding protein 6</td>\n",
       "      <td>BRK PF07533 2400-2441, Chromo PF00385 292-354 ...</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "      <td>Chromatin remodeling</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "      <td>chromatin</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>392</th>\n",
       "      <td>23411</td>\n",
       "      <td>SIR2L1</td>\n",
       "      <td>Eraser</td>\n",
       "      <td>sirtuin 1</td>\n",
       "      <td>Pfam-B_28511 PB028511 1-109, Pfam-B_5913 PB005...</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "      <td>Histone modification erase, Histone modificati...</td>\n",
       "      <td>Histone acetylation, Histone methylation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>393</th>\n",
       "      <td>6419</td>\n",
       "      <td>MARCHF1</td>\n",
       "      <td>Writer</td>\n",
       "      <td>membrane associated ring-CH-type finger 1</td>\n",
       "      <td>RINGv (PF12906)</td>\n",
       "      <td>RNF</td>\n",
       "      <td>Ring finger proteins</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394</th>\n",
       "      <td>6419</td>\n",
       "      <td>RNF171</td>\n",
       "      <td>Writer</td>\n",
       "      <td>membrane associated ring-CH-type finger 1</td>\n",
       "      <td>RINGv (PF12906)</td>\n",
       "      <td>RNF</td>\n",
       "      <td>Ring finger proteins</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395</th>\n",
       "      <td>6419</td>\n",
       "      <td>MARCH-I</td>\n",
       "      <td>Writer</td>\n",
       "      <td>membrane associated ring-CH-type finger 1</td>\n",
       "      <td>RINGv (PF12906)</td>\n",
       "      <td>RNF</td>\n",
       "      <td>Ring finger proteins</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>396</th>\n",
       "      <td>6419</td>\n",
       "      <td>FLJ20668</td>\n",
       "      <td>Writer</td>\n",
       "      <td>membrane associated ring-CH-type finger 1</td>\n",
       "      <td>RINGv (PF12906)</td>\n",
       "      <td>RNF</td>\n",
       "      <td>Ring finger proteins</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>397 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ID (REMOVE)    SYMBOL     REW  \\\n",
       "0          10389     SCML2  Reader   \n",
       "1          22955     SCMH1  Reader   \n",
       "2           5252      PHF1  Reader   \n",
       "3           8726       EED  Reader   \n",
       "4          84181      CHD6  Reader   \n",
       "..           ...       ...     ...   \n",
       "392        23411    SIR2L1  Eraser   \n",
       "393         6419   MARCHF1  Writer   \n",
       "394         6419    RNF171  Writer   \n",
       "395         6419   MARCH-I  Writer   \n",
       "396         6419  FLJ20668  Writer   \n",
       "\n",
       "                              HGNC approved name  \\\n",
       "0         sex comb on midleg-like 2 (Drosophila)   \n",
       "1      sex comb on midleg homolog 1 (Drosophila)   \n",
       "2                           PHD finger protein 1   \n",
       "3                 embryonic ectoderm development   \n",
       "4    chromodomain helicase DNA binding protein 6   \n",
       "..                                           ...   \n",
       "392                                    sirtuin 1   \n",
       "393    membrane associated ring-CH-type finger 1   \n",
       "394    membrane associated ring-CH-type finger 1   \n",
       "395    membrane associated ring-CH-type finger 1   \n",
       "396    membrane associated ring-CH-type finger 1   \n",
       "\n",
       "                                          Pfam domains HGNC gene family tag  \\\n",
       "0    DUF3588 PF12140 353-468, MBT PF02820 67-139 17...                 SAMD   \n",
       "1    DUF3588 PF12140 355-469, MBT PF02820 62-134 17...                 SAMD   \n",
       "2    Mtf2_C PF14061 525-566, PHD PF00628 89-142, Pf...            TDRD, PHF   \n",
       "3    Pfam-B_26038 PB026038 1-39, Pfam-B_6684 PB0066...                  WDR   \n",
       "4    BRK PF07533 2400-2441, Chromo PF00385 292-354 ...                    #   \n",
       "..                                                 ...                  ...   \n",
       "392  Pfam-B_28511 PB028511 1-109, Pfam-B_5913 PB005...                    #   \n",
       "393                                    RINGv (PF12906)                  RNF   \n",
       "394                                    RINGv (PF12906)                  RNF   \n",
       "395                                    RINGv (PF12906)                  RNF   \n",
       "396                                    RINGv (PF12906)                  RNF   \n",
       "\n",
       "                        HGNC gene family description  \\\n",
       "0        Sterile alpha motif (SAM) domain containing   \n",
       "1        Sterile alpha motif (SAM) domain containing   \n",
       "2    Tudor domain containing, Zinc fingers, PHD-type   \n",
       "3                        WD repeat domain containing   \n",
       "4                                                  #   \n",
       "..                                               ...   \n",
       "392                                                #   \n",
       "393                             Ring finger proteins   \n",
       "394                             Ring finger proteins   \n",
       "395                             Ring finger proteins   \n",
       "396                             Ring finger proteins   \n",
       "\n",
       "                                              Function  \\\n",
       "0                         Polycomb group (PcG) protein   \n",
       "1                         Polycomb group (PcG) protein   \n",
       "2                         Polycomb group (PcG) protein   \n",
       "3                         Polycomb group (PcG) protein   \n",
       "4                                 Chromatin remodeling   \n",
       "..                                                 ...   \n",
       "392  Histone modification erase, Histone modificati...   \n",
       "393                                                NaN   \n",
       "394                                                NaN   \n",
       "395                                                NaN   \n",
       "396                                                NaN   \n",
       "\n",
       "                                 Modification Protein complex Target molecule  \\\n",
       "0                                           #            PRC1               #   \n",
       "1                                           #            PRC1               #   \n",
       "2                                           #            PRC2               #   \n",
       "3                                           #            PRC2               #   \n",
       "4                                           #               #       chromatin   \n",
       "..                                        ...             ...             ...   \n",
       "392  Histone acetylation, Histone methylation             NaN             NaN   \n",
       "393                                       NaN             NaN             NaN   \n",
       "394                                       NaN             NaN             NaN   \n",
       "395                                       NaN             NaN             NaN   \n",
       "396                                       NaN             NaN             NaN   \n",
       "\n",
       "    Target entity Product  \n",
       "0               #       #  \n",
       "1               #       #  \n",
       "2               #       #  \n",
       "3               #       #  \n",
       "4               #       #  \n",
       "..            ...     ...  \n",
       "392           NaN     NaN  \n",
       "393           NaN     NaN  \n",
       "394           NaN     NaN  \n",
       "395           NaN     NaN  \n",
       "396           NaN     NaN  \n",
       "\n",
       "[397 rows x 13 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.replace(\"\\#\", np.nan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "id": "-SBpH30FtKay"
   },
   "outputs": [],
   "source": [
    "#Dropping columns 'ID' and 'SYMBOL' because they are not features\n",
    "data.drop(['ID (REMOVE)', 'SYMBOL', 'HGNC approved name'], axis = 1, inplace = True)\n",
    "data.replace('#', np.nan, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "b7Q6sAi8tQ1q",
    "outputId": "db267d29-2622-4f4c-c413-eb9456edc6d0"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final Shape of Dataset: (393, 10)\n"
     ]
    }
   ],
   "source": [
    "#Checking where all data in the columns are NaN\n",
    "to_check = data.columns[5:]\n",
    "data[to_check].isna().all(1).sum()\n",
    "data.dropna(how='all', subset = to_check, inplace = True)\n",
    "print('Final Shape of Dataset:', data.shape)\n",
    "data.replace(np.nan, '-', inplace = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "DPCLHahztTr_"
   },
   "source": [
    "### Categorical Encoding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "id": "CPsaNkOPuwVs",
    "outputId": "c85d8e2d-5151-4e8e-f9ab-b2c9cc312603"
   },
   "outputs": [],
   "source": [
    "X = data.iloc[:, 1:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "8bfp5dRLu2Bw",
    "outputId": "fd2f1e98-4ab5-42ab-ca7e-a23a9dfe67bf"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(393, 579)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.preprocessing import OneHotEncoder\n",
    "\n",
    "encoder = OneHotEncoder(handle_unknown='ignore', sparse = False)\n",
    "X_encoded = encoder.fit_transform(X)\n",
    "X_encoded.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Class Rebalancing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "id": "8NzJtNe0vk00"
   },
   "outputs": [],
   "source": [
    "from imblearn.over_sampling import SMOTE\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from collections import Counter\n",
    "\n",
    "y_df = data['REW']\n",
    "\n",
    "X = X_encoded\n",
    "y = y_df.values\n",
    "\n",
    "y_ = LabelEncoder().fit_transform(y)\n",
    "oversample = SMOTE(random_state=42)\n",
    "X_smote, y_smote = oversample.fit_resample(X, y_)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions for Tested Models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Naive Bayes Model\n",
    "from sklearn.model_selection import LeaveOneOut\n",
    "\n",
    "loocv = LeaveOneOut()\n",
    "\n",
    "model_cv =GaussianNB()\n",
    "\n",
    "predicted_prob_dict_cv = dict()\n",
    "auc_s_cv = []\n",
    "accuracy_s_cv = []\n",
    "f1_s_cv = []\n",
    "ytest_cv = []\n",
    "ypred_cv = []\n",
    "    \n",
    "#score_id\n",
    "scores_id_cv = []\n",
    "scores_cv = []\n",
    "sample_ids_cv = np.array(data.index)\n",
    "\n",
    "y_pred_proba_cv_list = []\n",
    "\n",
    "for fold, (train_idx, test_idx) in enumerate(loocv.split(X_smote)):   \n",
    "  X_train_cv = X_smote[train_idx]\n",
    "  X_test_cv = X_smote[test_idx]\n",
    "\n",
    "  anova_filter_cv = SelectKBest(f_classif, k= 8)\n",
    "  var_filter_cv = VarianceThreshold(threshold= .8 * (1 - .8))\n",
    "  anova_pipeline_cv = make_pipeline(var_filter_cv, anova_filter_cv, model)\n",
    "  anova_pipeline_cv.fit(X_train_cv, y_smote[train_idx])\n",
    "  y_pred_prob_cv = anova_pipeline_cv.predict_proba(X_test_cv)\n",
    "  y_pred_cv = anova_pipeline.predict(X_test_cv)\n",
    "  \n",
    "  \n",
    "  #auc_cv = roc_auc_score(y_smote[test_idx], y_pred_prob_cv.flatten(), multi_class='ovr')\n",
    "  y_pred_proba_cv_list.append(y_pred_prob_cv.flatten())\n",
    "  #auc_s_cv.append(auc_cv)\n",
    "  accuracy_s_cv.append(accuracy_score(y_smote[test_idx], y_pred_cv))\n",
    "  f1_s_cv.append(f1_score(y_smote[test_idx], y_pred_cv, average=\"weighted\"))\n",
    "  scores_id_cv.append(test_idx)\n",
    "  scores_cv.append(y_pred_prob_cv)\n",
    "\n",
    "\n",
    "scores_id_cv = np.concatenate(scores_id_cv, axis=0) \n",
    "scores_cv = np.concatenate(scores_cv, axis=0) \n",
    "map_ids = []\n",
    "for i in range(len(sample_ids_cv)):\n",
    "  map_ids.append(np.where(scores_id_cv==i)[0][0])\n",
    "map_scores = scores_cv[map_ids]\n",
    "y_predict_cv = np.argmax(map_scores, axis=1)\n",
    "\n",
    "confusion_mat = confusion_matrix(y_, y_predict_cv)\n",
    "\n",
    "sns.set(style=\"white\")\n",
    "labs=['Eraser','Writer', 'Reader']\n",
    "# Generate a large random dataset\n",
    "plt.figure(figsize=(4, 4))\n",
    "sns.set(font_scale=1.2)\n",
    "ax = sns.heatmap(confusion_mat, annot=True, square=True, xticklabels=labs, yticklabels=labs, cmap='RdPu', fmt=\"d\",cbar_kws={\"shrink\": 0.85})\n",
    "ax.xaxis.tick_top()\n",
    "ax.yaxis.tick_left()\n",
    "ax.set_xticklabels(labs, fontsize=14)\n",
    "ax.set_yticklabels(labs, fontsize=14)\n",
    "plt.xlabel('Naive Bayes', fontsize=14, labelpad=11)\n",
    "plt.yticks(rotation=0) \n",
    "plt.yticks(rotation=0) \n",
    "\n",
    "plt.savefig(\"nb_loocv\", format=\"png\",bbox_inches='tight', dpi=600)\n",
    "\n",
    "accuracy_cv = np.mean(accuracy_s_cv)\n",
    "f1_cv = np.mean(f1_s_cv)\n",
    "auc_Score_cv = roc_auc_score(y_smote, y_pred_proba_cv_list, multi_class='ovr')\n",
    "predicted_prob_dict_cv[fold] = y_pred_prob_cv\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Accuracy for Naive Bayes\" , accuracy_cv)\n",
    "print(\"F1-score for Naive Bayes\" , f1_cv)\n",
    "print(\"AUC score for Naive Bayes\" , auc_Score_cv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ANN Model\n",
    "from sklearn.neural_network import MLPClassifier\n",
    "from sklearn.model_selection import LeaveOneOut\n",
    "\n",
    "    \n",
    "auc_s_loocv = []\n",
    "accuracy_s_loocv = []\n",
    "f1_s_loocv = []\n",
    "    \n",
    "    #score_id\n",
    "scores_id_loocv = []\n",
    "scores_loocv = []\n",
    "sample_ids_loocv = np.array(data.index)\n",
    "            \n",
    "loocv = LeaveOneOut()\n",
    "\n",
    "clf = MLPClassifier(hidden_layer_sizes = (100,100,), random_state=0, max_iter=5000)\n",
    "    \n",
    "for fold, (train_idx, test_idx) in enumerate(loocv.split(X_smote)):\n",
    "  X_train = X_smote[train_idx]\n",
    "  X_test = X_smote[test_idx]\n",
    "  anova_filter = SelectKBest(f_classif, k= 9)\n",
    "  var_filter = VarianceThreshold(threshold= .8 * (1 - .8))\n",
    "  anova_knn = make_pipeline(var_filter, anova_filter, clf)\n",
    "  anova_knn.fit(X_train, y_smote[train_idx])\n",
    "  y_pred_prob = anova_knn.predict_proba(X_test)\n",
    "  y_pred = anova_knn.predict(X_test)\n",
    "\n",
    "  auc_s_loocv.append(y_pred_prob.flatten())\n",
    "  accuracy_s_loocv.append(accuracy_score(y_smote[test_idx], y_pred))\n",
    "  f1_s_loocv.append(f1_score(y_smote[test_idx], y_pred, average='weighted'))      \n",
    "  scores_id_loocv.append(test_idx)\n",
    "  scores_loocv.append(y_pred_prob)\n",
    "        \n",
    "scores_id_loocv = np.concatenate(scores_id_loocv, axis=0) \n",
    "scores_loocv = np.concatenate(scores_loocv, axis=0) \n",
    "map_ids_loocv = []\n",
    "for i in range(len(sample_ids_loocv)):\n",
    "    map_ids_loocv.append(np.where(scores_id_loocv==i)[0][0])\n",
    "    map_scores_loocv = scores_loocv[map_ids_loocv]\n",
    "y_predict_loocv = np.argmax(map_scores_loocv, axis=1)\n",
    "    \n",
    "confusion_mat = confusion_matrix(y_, y_predict_loocv)\n",
    "sns.set(style=\"white\")\n",
    "labs=['Eraser','Writer', 'Reader']\n",
    "    # Generate a large random dataset\n",
    "plt.figure(figsize=(4, 4))\n",
    "sns.set(font_scale=1.2)\n",
    "ax = sns.heatmap(confusion_mat, annot=True, square=True, xticklabels=labs, yticklabels=labs, cmap='RdPu', fmt=\"d\",cbar_kws={\"shrink\": 0.85})\n",
    "ax.xaxis.tick_top()\n",
    "ax.yaxis.tick_left()\n",
    "ax.set_xticklabels(labs, fontsize=14)\n",
    "ax.set_yticklabels(labs, fontsize=14)\n",
    "plt.xlabel('Deep Neural Net', fontsize=14, labelpad=11)\n",
    "\n",
    "\n",
    "plt.yticks(rotation=0) \n",
    "plt.xticks(rotation=0) \n",
    "plt.savefig(\"dnn_loocv\", format=\"png\",bbox_inches='tight', dpi=600)\n",
    "plt.show()\n",
    "            \n",
    "#avg_score = np.mean(auc_s)\n",
    "accuracy_ann = np.mean(accuracy_s_loocv)\n",
    "f1_ann = np.mean(f1_s_loocv)\n",
    "auc_ann = roc_auc_score(y_smote, auc_s_loocv, multi_class='ovr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Accuracy for ANN\" , accuracy_ann)\n",
    "print(\"F1-score for ANN\" , f1_ann)\n",
    "print(\"AUC score for ANN\" ,  roc_auc_score(y_smote, auc_s_loocv, multi_class='ovr'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#KNN Model\n",
    "import numpy as np\n",
    "from sklearn import preprocessing\n",
    "from sklearn.metrics import mean_squared_error\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.feature_selection import SelectKBest, f_classif\n",
    "from sklearn.pipeline import make_pipeline\n",
    "from sklearn.feature_selection import VarianceThreshold\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.metrics import f1_score\n",
    "from sklearn.metrics import precision_recall_curve\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "def knn_kfolds(X, y, n_folds, n_neighbors, new_data, old_y, random_state=None):\n",
    "    \n",
    "    auc_s = []\n",
    "    accuracy_s = []\n",
    "    f1_s = []\n",
    "    ytest = []\n",
    "    ypred = []\n",
    "    \n",
    "    scores_id = []\n",
    "    scores = []\n",
    "    sample_ids = np.array(new_data.index)\n",
    "    \n",
    "    kf = StratifiedKFold(n_splits=n_folds, shuffle=True, random_state = random_state)\n",
    "    \n",
    "    model = KNeighborsClassifier(n_neighbors= n_neighbors)\n",
    "    \n",
    "    loocv = LeaveOneOut()\n",
    "    y_pred_proba_cv_list = []\n",
    "    \n",
    "    \n",
    "    for fold, (train_idx, test_idx) in enumerate(loocv.split(X,y)):\n",
    "        \n",
    "        scaler = preprocessing.StandardScaler().fit(X)\n",
    "        X_train_scaled = scaler.transform(X[train_idx])\n",
    "        X_test_scaled = scaler.transform(X[test_idx])\n",
    "        \n",
    "        anova_filter = SelectKBest(f_classif, k= 9)\n",
    "        var_filter = VarianceThreshold(threshold= .8 * (1 - .8))\n",
    "        anova_knn = make_pipeline(var_filter, anova_filter, model)\n",
    "        anova_knn.fit(X_train_scaled, y[train_idx])\n",
    "        y_pred_prob = anova_knn.predict_proba(X_test_scaled)\n",
    "        y_pred = anova_knn.predict(X_test_scaled)\n",
    "        y_pred_proba_cv_list.append(y_pred_prob.flatten())\n",
    "        \n",
    "        \n",
    "        accuracy_s.append(accuracy_score(y[test_idx], y_pred))\n",
    "        f1_s.append(f1_score(y[test_idx], y_pred, average=\"weighted\"))\n",
    "        scores_id.append(test_idx)\n",
    "        scores.append(y_pred_prob)\n",
    "        \n",
    "    scores_id = np.concatenate(scores_id, axis=0) \n",
    "    scores = np.concatenate(scores, axis=0) \n",
    "    map_ids = []\n",
    "    for i in range(len(sample_ids)):\n",
    "        map_ids.append(np.where(scores_id==i)[0][0])\n",
    "    map_scores = scores[map_ids]\n",
    "    y_predict = np.argmax(map_scores, axis=1)\n",
    "    \n",
    "    confusion_mat = confusion_matrix(old_y, y_predict)\n",
    "\n",
    "    sns.set(style=\"white\")\n",
    "    labs=['Eraser','Writer', 'Reader']\n",
    "    \n",
    "    # Generate a large random dataset\n",
    "    plt.figure(figsize=(4, 4))\n",
    "    sns.set(font_scale=1.2)\n",
    "    ax = sns.heatmap(confusion_mat, annot=True, square=True, xticklabels=labs, yticklabels=labs, cmap='RdPu', fmt=\"d\",cbar_kws={\"shrink\": 0.85})\n",
    "    ax.xaxis.tick_top()\n",
    "    ax.yaxis.tick_left()\n",
    "    ax.set_xticklabels(labs, fontsize=14)\n",
    "    ax.set_yticklabels(labs, fontsize=14)\n",
    "    plt.xlabel('K-Neighbors Classifier', fontsize=14, labelpad=11)\n",
    "\n",
    "    plt.yticks(rotation=0) \n",
    "    plt.xticks(rotation=0) \n",
    "    plt.show()\n",
    "    \n",
    "    avg_score = roc_auc_score(y_smote, y_pred_proba_cv_list, multi_class='ovr')\n",
    "    accuracy = np.mean(accuracy_s)\n",
    "    f1 = np.mean(f1_s)\n",
    "\n",
    "    return [avg_score, accuracy, f1, confusion_mat]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Decision Tree Model\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "\n",
    "def decision_tree(X, y, n_folds, max_depth, new_data, old_y, random_state=None):\n",
    "    \n",
    "    auc_s = [] \n",
    "    accuracy_s = []\n",
    "    f1_s = []\n",
    "    \n",
    "    #score_id\n",
    "    scores_id = []\n",
    "    scores = []\n",
    "    sample_ids = np.array(new_data.index)\n",
    "            \n",
    "    kf = StratifiedKFold(n_splits=n_folds, shuffle=True, random_state = random_state)\n",
    "\n",
    "    model = DecisionTreeClassifier(max_depth= max_depth)\n",
    "    \n",
    "    loocv = LeaveOneOut()\n",
    "    y_pred_proba_cv_list = []\n",
    "    \n",
    "    for fold, (train_idx, test_idx) in enumerate(loocv.split(X,y)):\n",
    "        \n",
    "        scaler = preprocessing.StandardScaler().fit(X)\n",
    "        X_train_scaled = scaler.transform(X[train_idx])\n",
    "        X_test_scaled = scaler.transform(X[test_idx])\n",
    "        \n",
    " \n",
    "        anova_filter = SelectKBest(f_classif, k= 9)\n",
    "        var_filter = VarianceThreshold(threshold=(.8 * (1 - .8)))\n",
    "        anova_knn = make_pipeline(anova_filter, model)\n",
    "        anova_knn.fit(X_train_scaled, y[train_idx])\n",
    "        y_pred_prob = anova_knn.predict_proba(X_test_scaled)\n",
    "        y_pred = anova_knn.predict(X_test_scaled)\n",
    "        \n",
    "        y_pred_proba_cv_list.append(y_pred_prob.flatten())\n",
    "\n",
    "        accuracy_s.append(accuracy_score(y[test_idx], y_pred))\n",
    "        f1_s.append(f1_score(y[test_idx], y_pred, average='weighted'))\n",
    "     \n",
    "        scores_id.append(test_idx)\n",
    "        scores.append(y_pred_prob)\n",
    "        \n",
    "    scores_id = np.concatenate(scores_id, axis=0) \n",
    "    scores = np.concatenate(scores, axis=0) \n",
    "    map_ids = []\n",
    "    for i in range(len(sample_ids)):\n",
    "        map_ids.append(np.where(scores_id==i)[0][0])\n",
    "    map_scores = scores[map_ids]\n",
    "    y_predict = np.argmax(map_scores, axis=1)\n",
    "    \n",
    "    confusion_mat = confusion_matrix(old_y, y_predict)\n",
    "\n",
    "    sns.set(style=\"white\")\n",
    "    labs=['Eraser','Writer', 'Reader']\n",
    "    \n",
    "    # Generate a large random dataset\n",
    "    plt.figure(figsize=(4, 4))\n",
    "    sns.set(font_scale=1.2)\n",
    "    ax = sns.heatmap(confusion_mat, annot=True, square=True, xticklabels=labs, yticklabels=labs, cmap='RdPu', fmt=\"d\",cbar_kws={\"shrink\": 0.85})\n",
    "    ax.xaxis.tick_top()\n",
    "    ax.yaxis.tick_left()\n",
    "    ax.set_xticklabels(labs, fontsize=14)\n",
    "    ax.set_yticklabels(labs, fontsize=14)\n",
    "    plt.xlabel('Decision Tree', fontsize=14, labelpad=11)\n",
    "\n",
    "    plt.yticks(rotation=0) \n",
    "    plt.xticks(rotation=0) \n",
    "    plt.show()\n",
    "        \n",
    "    avg_score = roc_auc_score(y_smote, y_pred_proba_cv_list, multi_class='ovr')\n",
    "    accuracy = np.mean(accuracy_s)\n",
    "    f1 = np.mean(f1_s)\n",
    "\n",
    "    return [avg_score, accuracy, f1, confusion_mat]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Random Forest Model\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "\n",
    "def random_forest(X, y, n_folds, max_depth, new_data, old_y, random_state=None):\n",
    "    \n",
    "    auc_s = []\n",
    "    accuracy_s = []\n",
    "    f1_s = []\n",
    "    \n",
    "    #score_id\n",
    "    scores_id = []\n",
    "    scores = []\n",
    "    sample_ids = np.array(new_data.index)\n",
    "            \n",
    "    kf = StratifiedKFold(n_splits=n_folds, shuffle=True, random_state = random_state)\n",
    "\n",
    "    model = RandomForestClassifier(max_depth= max_depth)\n",
    "    \n",
    "    loocv = LeaveOneOut()\n",
    "    y_pred_proba_cv_list = []\n",
    "    \n",
    "    for fold, (train_idx, test_idx) in enumerate(loocv.split(X,y)):\n",
    "        \n",
    "        scaler = preprocessing.StandardScaler().fit(X)\n",
    "        X_train_scaled = scaler.transform(X[train_idx])\n",
    "        X_test_scaled = scaler.transform(X[test_idx])\n",
    "        \n",
    " \n",
    "        anova_filter = SelectKBest(f_classif, k= 9)\n",
    "        var_filter = VarianceThreshold(threshold= .8 * (1 - .8))\n",
    "        anova_knn = make_pipeline(var_filter, anova_filter, model)\n",
    "        anova_knn.fit(X_train_scaled, y[train_idx])\n",
    "        y_pred_prob = anova_knn.predict_proba(X_test_scaled)\n",
    "        y_pred = anova_knn.predict(X_test_scaled)\n",
    "\n",
    "        y_pred_proba_cv_list.append(y_pred_prob.flatten())\n",
    "\n",
    "        accuracy_s.append(accuracy_score(y[test_idx], y_pred))\n",
    "        f1_s.append(f1_score(y[test_idx], y_pred, average='weighted'))\n",
    "        \n",
    "        \n",
    "        scores_id.append(test_idx)\n",
    "        scores.append(y_pred_prob)\n",
    "        \n",
    "    scores_id = np.concatenate(scores_id, axis=0) \n",
    "    scores = np.concatenate(scores, axis=0) \n",
    "    map_ids = []\n",
    "    for i in range(len(sample_ids)):\n",
    "        map_ids.append(np.where(scores_id==i)[0][0])\n",
    "    map_scores = scores[map_ids]\n",
    "    y_predict = np.argmax(map_scores, axis=1)\n",
    "    \n",
    "    confusion_mat = confusion_matrix(old_y, y_predict)\n",
    "\n",
    "    sns.set(style=\"white\")\n",
    "    labs=['Eraser','Writer', 'Reader']\n",
    "    \n",
    "    # Generate a large random dataset\n",
    "    plt.figure(figsize=(4, 4))\n",
    "    sns.set(font_scale=1.2)\n",
    "    ax = sns.heatmap(confusion_mat, annot=True, square=True, xticklabels=labs, yticklabels=labs, cmap='RdPu', fmt=\"d\",cbar_kws={\"shrink\": 0.85})\n",
    "    ax.xaxis.tick_top()\n",
    "    ax.yaxis.tick_left()\n",
    "    ax.set_xticklabels(labs, fontsize=14)\n",
    "    ax.set_yticklabels(labs, fontsize=14)\n",
    "    plt.xlabel('Random Forest', fontsize=14, labelpad=11)\n",
    "\n",
    "    plt.yticks(rotation=0) \n",
    "    plt.xticks(rotation=0) \n",
    "    plt.show()\n",
    "        \n",
    "    avg_score = roc_auc_score(y_smote, y_pred_proba_cv_list, multi_class='ovr')\n",
    "    accuracy = np.mean(accuracy_s)\n",
    "    f1 = np.mean(f1_s)\n",
    "\n",
    "            \n",
    "    return [avg_score, accuracy, f1, confusion_mat]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#SVM Model \n",
    "\n",
    "from sklearn.pipeline import make_pipeline\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.svm import SVC\n",
    "\n",
    "def svm_f(X, y, n_folds, new_data, old_y, random_state=None):\n",
    "    \n",
    "    auc_s = []\n",
    "    accuracy_s = []\n",
    "    f1_s = []\n",
    "    \n",
    "    #score_id\n",
    "    scores_id = []\n",
    "    scores = []\n",
    "    sample_ids = np.array(new_data.index)\n",
    "    \n",
    "    kf = StratifiedKFold(n_splits=n_folds, shuffle=True, random_state = random_state)\n",
    "\n",
    "    clf = make_pipeline(StandardScaler(), SVC(gamma='auto', probability=True))\n",
    "    \n",
    "    loocv = LeaveOneOut()\n",
    "    y_pred_proba_cv_list = []\n",
    "    \n",
    "    for fold, (train_idx, test_idx) in enumerate(loocv.split(X,y)):\n",
    "        \n",
    "        scaler = preprocessing.StandardScaler().fit(X)\n",
    "        X_train_scaled = scaler.transform(X[train_idx])\n",
    "        X_test_scaled = scaler.transform(X[test_idx])\n",
    "        \n",
    " \n",
    "        anova_filter = SelectKBest(f_classif, k= 9)\n",
    "        var_filter = VarianceThreshold(threshold= .8 * (1 - .8))\n",
    "        anova_knn = make_pipeline(var_filter, anova_filter, clf)\n",
    "        anova_knn.fit(X_train_scaled, y[train_idx])\n",
    "        y_pred_prob = anova_knn.predict_proba(X_test_scaled)\n",
    "        y_pred = anova_knn.predict(X_test_scaled)\n",
    "        \n",
    "        y_pred_proba_cv_list.append(y_pred_prob.flatten())\n",
    "\n",
    "        accuracy_s.append(accuracy_score(y[test_idx], y_pred))\n",
    "        f1_s.append(f1_score(y[test_idx], y_pred, average='weighted'))\n",
    "        \n",
    "        scores_id.append(test_idx)\n",
    "        scores.append(y_pred_prob)\n",
    "        \n",
    "    scores_id = np.concatenate(scores_id, axis=0) \n",
    "    scores = np.concatenate(scores, axis=0) \n",
    "    map_ids = []\n",
    "    for i in range(len(sample_ids)):\n",
    "        map_ids.append(np.where(scores_id==i)[0][0])\n",
    "    map_scores = scores[map_ids]\n",
    "    y_predict = np.argmax(map_scores, axis=1)\n",
    "    \n",
    "    confusion_mat = confusion_matrix(old_y, y_predict)\n",
    "\n",
    "    sns.set(style=\"white\")\n",
    "    labs=['Eraser','Writer', 'Reader']\n",
    "    # Generate a large random dataset\n",
    "    plt.figure(figsize=(4, 4))\n",
    "    sns.set(font_scale=1.2)\n",
    "    ax = sns.heatmap(confusion_mat, annot=True, square=True, xticklabels=labs, yticklabels=labs, cmap='RdPu', fmt=\"d\",cbar_kws={\"shrink\": 0.85})\n",
    "    ax.xaxis.tick_top()\n",
    "    ax.yaxis.tick_left()\n",
    "    ax.set_xticklabels(labs, fontsize=14)\n",
    "    ax.set_yticklabels(labs, fontsize=14)\n",
    "    plt.xlabel('Support Vector Machine', fontsize=14, labelpad=11)\n",
    "\n",
    "    plt.yticks(rotation=0) \n",
    "    plt.xticks(rotation=0) \n",
    "    plt.show()\n",
    "            \n",
    "    avg_score = roc_auc_score(y_smote, y_pred_proba_cv_list, multi_class='ovr')\n",
    "    accuracy = np.mean(accuracy_s)\n",
    "    f1 = np.mean(f1_s)\n",
    "\n",
    "    return [avg_score, accuracy, f1, confusion_mat]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Feed-Forward Neural Network Model\n",
    "from sklearn.neural_network import MLPClassifier\n",
    "\n",
    "def fnn_f(X, y, n_folds, new_data, old_y, random_state=None):\n",
    "    \n",
    "    auc_s = []\n",
    "    accuracy_s = []\n",
    "    f1_s = []\n",
    "    \n",
    "    #score_id\n",
    "    scores_id = []\n",
    "    scores = []\n",
    "    sample_ids = np.array(new_data.index)\n",
    "            \n",
    "    kf = StratifiedKFold(n_splits=n_folds, shuffle=True, random_state = random_state)\n",
    "\n",
    "    clf = MLPClassifier(random_state=0, max_iter=5000)\n",
    "    \n",
    "    loocv = LeaveOneOut()\n",
    "    y_pred_proba_cv_list = []\n",
    "    \n",
    "    for fold, (train_idx, test_idx) in enumerate(loocv.split(X,y)):\n",
    "        \n",
    "        scaler = preprocessing.StandardScaler().fit(X)\n",
    "        X_train_scaled = scaler.transform(X[train_idx])\n",
    "        X_test_scaled = scaler.transform(X[test_idx])\n",
    "        \n",
    " \n",
    "        anova_filter = SelectKBest(f_classif, k= 9)\n",
    "        var_filter = VarianceThreshold(threshold= .8 * (1 - .8))\n",
    "        anova_knn = make_pipeline(var_filter, anova_filter, clf)\n",
    "        anova_knn.fit(X_train_scaled, y[train_idx])\n",
    "        y_pred_prob = anova_knn.predict_proba(X_test_scaled)\n",
    "        y_pred = anova_knn.predict(X_test_scaled)\n",
    "\n",
    "        y_pred_proba_cv_list.append(y_pred_prob.flatten())\n",
    "\n",
    "        accuracy_s.append(accuracy_score(y[test_idx], y_pred))\n",
    "        f1_s.append(f1_score(y[test_idx], y_pred, average='weighted'))\n",
    "        \n",
    "        scores_id.append(test_idx)\n",
    "        scores.append(y_pred_prob)\n",
    "        \n",
    "    scores_id = np.concatenate(scores_id, axis=0) \n",
    "    scores = np.concatenate(scores, axis=0) \n",
    "    map_ids = []\n",
    "    for i in range(len(sample_ids)):\n",
    "        map_ids.append(np.where(scores_id==i)[0][0])\n",
    "    map_scores = scores[map_ids]\n",
    "    y_predict = np.argmax(map_scores, axis=1)\n",
    "    \n",
    "    confusion_mat = confusion_matrix(old_y, y_predict)\n",
    "\n",
    "    sns.set(style=\"white\")\n",
    "    labs=['Eraser','Writer', 'Reader']\n",
    "    # Generate a large random dataset\n",
    "    plt.figure(figsize=(4, 4))\n",
    "    sns.set(font_scale=1.2)\n",
    "    ax = sns.heatmap(confusion_mat, annot=True, square=True, xticklabels=labs, yticklabels=labs, cmap='RdPu', fmt=\"d\",cbar_kws={\"shrink\": 0.85})\n",
    "    ax.xaxis.tick_top()\n",
    "    ax.yaxis.tick_left()\n",
    "    ax.set_xticklabels(labs, fontsize=14)\n",
    "    ax.set_yticklabels(labs, fontsize=14)\n",
    "    plt.xlabel('Feed-Forward Neural Network', fontsize=14, labelpad=11)\n",
    "\n",
    "    plt.yticks(rotation=0) \n",
    "    plt.xticks(rotation=0) \n",
    "    plt.show()\n",
    "            \n",
    "    avg_score = roc_auc_score(y_smote, y_pred_proba_cv_list, multi_class='ovr')\n",
    "    accuracy = np.mean(accuracy_s)\n",
    "    f1 = np.mean(f1_s)\n",
    "\n",
    "    return [avg_score, accuracy, f1, confusion_mat]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Logistic Regression Model\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "\n",
    "def lr(X, y, n_folds, new_data, old_y, random_state=None):\n",
    "    \n",
    "    auc_s = []\n",
    "    accuracy_s = []\n",
    "    f1_s = []\n",
    "            \n",
    "    kf = StratifiedKFold(n_splits=n_folds, shuffle=True, random_state = random_state)\n",
    "\n",
    "    clf = LogisticRegression(penalty ='l2')\n",
    "    \n",
    "    #score_id\n",
    "    scores_id = []\n",
    "    scores = []\n",
    "    sample_ids = np.array(new_data.index)\n",
    "    \n",
    "    loocv = LeaveOneOut()\n",
    "    y_pred_proba_cv_list = []\n",
    "    \n",
    "    for fold, (train_idx, test_idx) in enumerate(loocv.split(X,y)):\n",
    "        \n",
    "        scaler = preprocessing.StandardScaler().fit(X)\n",
    "        X_train_scaled = scaler.transform(X[train_idx])\n",
    "        X_test_scaled = scaler.transform(X[test_idx])\n",
    "        \n",
    " \n",
    "        anova_filter = SelectKBest(f_classif, k= 9)\n",
    "        var_filter = VarianceThreshold(threshold= .8 * (1 - .8))\n",
    "        anova_knn = make_pipeline(var_filter, anova_filter, clf)\n",
    "        anova_knn.fit(X_train_scaled, y[train_idx])\n",
    "        y_pred_prob = anova_knn.predict_proba(X_test_scaled)\n",
    "        y_pred = anova_knn.predict(X_test_scaled)\n",
    "\n",
    "        y_pred_proba_cv_list.append(y_pred_prob.flatten())\n",
    "\n",
    "        accuracy_s.append(accuracy_score(y[test_idx], y_pred))\n",
    "        f1_s.append(f1_score(y[test_idx], y_pred, average='weighted'))\n",
    "        \n",
    "        scores_id.append(test_idx)\n",
    "        scores.append(y_pred_prob)\n",
    "        \n",
    "    scores_id = np.concatenate(scores_id, axis=0) \n",
    "    scores = np.concatenate(scores, axis=0) \n",
    "    map_ids = []\n",
    "    for i in range(len(sample_ids)):\n",
    "        map_ids.append(np.where(scores_id==i)[0][0])\n",
    "    map_scores = scores[map_ids]\n",
    "    y_predict = np.argmax(map_scores, axis=1)\n",
    "    \n",
    "    confusion_mat = confusion_matrix(old_y, y_predict)\n",
    "\n",
    "    sns.set(style=\"white\")\n",
    "    labs=['Eraser','Writer', 'Reader']\n",
    "    # Generate a large random dataset\n",
    "    plt.figure(figsize=(4, 4))\n",
    "    sns.set(font_scale=1.2)\n",
    "    ax = sns.heatmap(confusion_mat, annot=True, square=True, xticklabels=labs, yticklabels=labs, cmap='RdPu', fmt=\"d\",cbar_kws={\"shrink\": 0.85})\n",
    "    ax.xaxis.tick_top()\n",
    "    ax.yaxis.tick_left()\n",
    "    ax.set_xticklabels(labs, fontsize=14)\n",
    "    ax.set_yticklabels(labs, fontsize=14)\n",
    "    plt.xlabel('Logistic Regression', fontsize=14, labelpad=11)\n",
    "\n",
    "    plt.yticks(rotation=0) \n",
    "    plt.xticks(rotation=0) \n",
    "    plt.show()\n",
    "            \n",
    "    avg_score = roc_auc_score(y_smote, y_pred_proba_cv_list, multi_class='ovr')\n",
    "    accuracy = np.mean(accuracy_s)\n",
    "    f1 = np.mean(f1_s)\n",
    "\n",
    "            \n",
    "    return [avg_score, accuracy, f1, confusion_mat]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Model Evaluation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "xUUOcql0O25c"
   },
   "outputs": [],
   "source": [
    "#Running the ML models on the generated dataset\n",
    "\n",
    "from sklearn.metrics import plot_confusion_matrix\n",
    "from sklearn.metrics import confusion_matrix\n",
    "\n",
    "results = {}\n",
    "auc = []\n",
    "accuracy = []\n",
    "f1 = []\n",
    "confusion_matrices = []\n",
    "\n",
    "knn_values = []\n",
    "dt_values = []\n",
    "rf_values = []\n",
    "\n",
    "k = range(1,10)\n",
    "for i in k:\n",
    "    knn = knn_kfolds(X_smote, y_smote, 160, i, data, y_, random_state=5)\n",
    "    dt = decision_tree(X_smote, y_smote, 160, i, data, y_, random_state=5)\n",
    "    rf = random_forest(X_smote, y_smote, 160, i, data, y_, random_state=5) \n",
    "\n",
    "    knn_values.append(knn[0])\n",
    "    dt_values.append(dt[0])\n",
    "    rf_values.append(rf[0])\n",
    "\n",
    "svm = svm_f(X_smote, y_smote, 160, data, y_, random_state=5)\n",
    "fnn = fnn_f(X_smote, y_smote, 160, data, y_, random_state=5)\n",
    "lr_s = lr(X_smote, y_smote, 160, data, y_, random_state=5)\n",
    "\n",
    "auc.append(max(knn_values))\n",
    "auc.append(max(rf_values))\n",
    "auc.append(max(dt_values))\n",
    "auc.append(fnn[0])\n",
    "auc.append(lr_s[0])\n",
    "auc.append(svm[0])\n",
    "\n",
    "accuracy.append(knn[1])\n",
    "accuracy.append(dt[1])\n",
    "accuracy.append(rf[1])\n",
    "accuracy.append(fnn[1])\n",
    "accuracy.append(lr_s[1])\n",
    "accuracy.append(svm[1])\n",
    "\n",
    "f1.append(knn[2])\n",
    "f1.append(dt[2])\n",
    "f1.append(rf[2])\n",
    "f1.append(fnn[2])\n",
    "f1.append(lr_s[2])\n",
    "f1.append(svm[2])\n",
    "\n",
    "\n",
    "confusion_matrices.append(knn[3])\n",
    "confusion_matrices.append(dt[3])\n",
    "confusion_matrices.append(rf[3])\n",
    "confusion_matrices.append(fnn[3])\n",
    "confusion_matrices.append(lr_s[3])\n",
    "confusion_matrices.append(svm[3])\n",
    "\n",
    "results['auc'] = auc\n",
    "#results['accuracy'] = accuracy\n",
    "results['f1'] = f1\n",
    "results['accuracy'] = accuracy\n",
    "results['confusion_matrix'] = confusion_matrices"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plotting Confusion Matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plotting the generated confusion matrices from cell above\n",
    "\n",
    "sns.set(style=\"white\")\n",
    "labs=['Eraser','Writer', 'Reader']\n",
    "models = [\"K-Neighbors Classifier\", \"Decision Tree\", \"Random Forest\", \"Support Vector Machine\", \"Feed-Forward Neural Network\", \"Logistic Regression\"]\n",
    "i = 0\n",
    "img_file =\"loocv\"\n",
    "for m in results['confusion_matrix']:\n",
    "    plt.figure(figsize=(4, 4))\n",
    "    sns.set(font_scale=1.2)\n",
    "    ax = sns.heatmap(m, annot=True, square=True, xticklabels=labs, yticklabels=labs, cmap='RdPu', fmt=\"d\",cbar_kws={\"shrink\": 0.85})\n",
    "    ax.xaxis.tick_top()\n",
    "    ax.yaxis.tick_left()\n",
    "    ax.set_xticklabels(labs, fontsize=14)\n",
    "    ax.set_yticklabels(labs, fontsize=14)\n",
    "    plt.xlabel(models[i], fontsize=14, labelpad=11)\n",
    "\n",
    "    plt.yticks(rotation=0) \n",
    "    plt.xticks(rotation=0) \n",
    "    plt.savefig(img_file + str(i), format=\"png\",bbox_inches='tight', dpi=600)\n",
    "    i = i + 1"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
